src/zarr/buffer.py

```diff
@@ -436,6 +442,311 @@ def as_numpy_array_wrapper(
     return prototype.buffer.from_bytes(func(buf.as_numpy_array()))
 
 
+class GpuBuffer(Buffer):
+    """A flat contiguous memory block on the GPU
+
+    We use Buffer throughout Zarr to represent a contiguous block of memory.
+
+    A Buffer is backed by a underlying array-like instance that represents
+    the memory. The memory type is unspecified; can be regular host memory,
+    CUDA device memory, or something else. The only requirement is that the
+    array-like instance can be copied/converted to a regular Numpy array
+    (host memory).
+
+    Note
+    ----
+    This buffer is untyped, so all indexing and sizes are in bytes.
+
+    Parameters
+    ----------
+    array_like
+        array-like object that must be 1-dim, contiguous, and byte dtype.
+    """
+
+    def __init__(self, array_like: ArrayLike):
+        if cp is None:
+            raise ImportError("Cannot use GpuBuffer without cupy. Please install cupy.")
+
+        if array_like.ndim != 1:
+            raise ValueError("array_like: only 1-dim allowed")
+        if array_like.dtype != np.dtype("b"):
+            raise ValueError("array_like: only byte dtype allowed")
+
+        # if type(array_like) == cp.ndarray
+        #     self._data = array_like
+        if hasattr(array_like, "__cuda_array_interface__"):
+            self._data = cp.asarray(array_like)
+        else:
+            # Slow copy based path for arrays that don't support the __cuda_array_interface__
+            # TODO: Add a fast zero-copy path for arrays that support the dlpack protocol
+            msg = (
+                "Creating a GpuBuffer with an array that does not support the "
+                "__cuda_array_interface__ for zero-copy transfers, "
+                "falling back to slow copy based path"
+            )
+            warnings.warn(
+                msg,
+                stacklevel=2,
+            )
+            buffer = Buffer(array_like)
+            self._data = cp.asarray(buffer.as_numpy_array())
+
+    @classmethod
+    def create_zero_length(cls) -> Self:
+        """Create an empty buffer with length zero
+
+        Returns
+        -------
+            New empty 0-length buffer
+        """
+        return cls(cp.array([], dtype="b"))
+
+    @classmethod
+    def from_buffer(cls, buffer: Buffer) -> Self:
+        """Create an GpuBuffer given an arbitrary Buffer
+
+        Returns
+        -------
+            New empty 0-length buffer
+        """
+        return cls(buffer.as_array_like())
+
+    @classmethod
+    def from_array_like(cls, array_like: ArrayLike) -> Self:
+        """Create a new buffer of a array-like object
+
+        Parameters
+        ----------
+        array_like
+            array-like object that must be 1-dim, contiguous, and byte dtype.
+
+        Returns
+        -------
+            New buffer representing `array_like`
+        """
+        return cls(array_like)
```

Comment:
Agree with this for `from_array_like`. However, `from_buffer` is not in the parent `Buffer` class. Maybe good to move this to the parent class then but have it default to go through numpy arrays?
```python
    @classmethod
    def from_buffer(cls, buffer: Buffer) -> Self:
        """Create a Buffer given another arbitrary Buffer
        """
        return cls(buffer.as_numpy_array())
```

Reply:
Sorry, I missed the previous PR that added `from_buffer` to the parent class. However, I feel like the `Buffer` class' `from_buffer` should use `as_numpy_array` instead of `as_array_like`. Otherwise, if you use the `GpuMemoryStore` or any other component that outputs a `GpuBuffer` and you try to coerce it to a regular buffer using `default_prototype.buffer.from_buffer(gpu_buffer)`, that will create a new `Buffer` object with `_data` being a cupy array. A downstream CPU compressor will then call `.as_numpy_array` on this Buffer class and will see an error similar to the snippet below right?
```python
>>> np.asanyarray(cp.random.random(5))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 np.asanyarray(cp.random.random(5))

File cupy/_core/core.pyx:1479, in cupy._core.core._ndarray_base.__array__()

TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.

```

Reply:
Good point, I guess the underlying problem is that `Buffer` requires that `ArrayLike` supports `np.asanyarray()`.

We might consider adding some some kind of `to-numpy-function/-method` to the `ArrayLike` protocol in a future PR.

Reply:
Agreed, we can add some functionality like that in the future. For now, I'll switch the base class to use `.as_numpy_array` which is equivalent here but allows interop. Also added a docstring note that subclassed of `Buffer` should implement more optimal `from_buffer` methods that can avoid copying when possible.

Reply:
Looks good